docs/architecture/adr-009-modular-antehandler.md

```diff
@@ -0,0 +1,173 @@
+# ADR 9: Modular AnteHandler
+
+## Changelog
+
+- 2019 Aug 31: Initial draft
+
+## Context
+
+The current AnteHandler design allows users to either use the default AnteHandler provided in `x/auth` or to build their own AnteHandler from scratch. Ideally AnteHandler functionality is split into multiple, modular functions that can be chained together along with custom ante-functions so that users do not have to rewrite common antehandler logic when they want to implement custom behavior.
+
+## Proposals
+
+### Per-Module AnteHandler
+
+One approach is to use the module manager and have each module implement its own antehandler if it requires custom antehandler logic. The ModuleManager can then be passed in an AnteHandler order in the same way it has an order for BeginBlockers and EndBlockers. The ModuleManager returns a single AnteHandler function that will take in a tx and run each module's antehandler in the specified order. The module manager's AnteHandler is set as the baseapp's AnteHandler.
+
+Pros:
+1. Simple to implement
+2. Can reuse ModuleManager architecture
+
+Cons:
+1. Improves granularity but still cannot get more granular than a per-module basis. If auth's `AnteHandle` function is in charge of validating memo and signatures, users cannot swap the signature-checking functionality while keeping the rest of auth's AnteHandle functionality.
+2. Module AnteHandler are run one after the other. There is no way for one AnteHandler to wrap or "decorate" another.
+
+### Weave Decorators
+
+The Weave approach acheives modularity through the use of Decorators. Weave introduces a new interface:
+
+```golang
+// Decorator wraps a Handler to provide common functionality
+// like authentication, or fee-handling, to many Handlers
+type Decorator interface {
+	Check(ctx Context, store KVStore, tx Tx, next Checker) (*CheckResult, error)
+	Deliver(ctx Context, store KVStore, tx Tx, next Deliverer) (*DeliverResult, error)
+}
+```
+
+Each decorator works like a modularized SDK antehandler function, but it can take in a `next` argument that may be another decorator or a Handler (which does not take in a next argument). These decorators can be chained together, one decorator being passed in as the `next` argument of the previous decorator in the chain. The chain ends in a Router which can take a tx and route to the appropriate msg handler.
+
+A key benefit of this approach is that one Decorator can wrap its internal logic around the next Checker/Deliverer. A weave Decorator may do the following:
```

Comment:
One example of where the decorator approach would be useful for us is to have a Decorator that wraps all other AnteHandlers with a defer clause that will catch an `OutOfGas` panic and return the appropriate sdk Result and abort.

In the per-module implementation, it was clear that having a defer clause in every antehandler made no sense. The best way I found to handle this was to set the `ctx.GasMeter` in the module manager's antehandler and then place a defer clause there. However, doing this required changing the `sdk.Tx` interface to include a `Gas` method. 
This can be seen [here](https://github.com/cosmos/cosmos-sdk/blob/aditya/modular-ante/types/module/module.go#L299).

Having a decorator that wraps all subsequent antehandler functions to set the `GasMeter` and implements the defer clause correctly before calling `next` may help us avoid changing the `sdk.Tx` interface. However, any chain that uses something other than `auth.StdTx` will have to implement their own decorator.